### DISCUSSION TITLE
Document filtering/querying with multiple key value pairs

### DISCUSSION THREAD

**Discussion by Jooseba** [ravendb/ravendb#17316] [Q&A]
Hi,

I need to filter documents by a long list of value key pairs. For now I have been using async sessions and querying them one by one but I was wondering is there a better way to do this? Can I somehow pass or map the key value pairs to filter/query so I could to that in one query instead of multiple queries in foreach loop?

```cs
            var multiTask = Parallel.ForEachAsync(dictionary, async (kvp, cancelToken) =>
            {
                    using (var session = DSHolder.Store.OpenAsyncSession(AppSettings.LayoutDb))
                    {
                        var result = await session.Query<TResult>().Where(x => x.Id == kvp.Key && x.OtherField == kvp.Value)).SingleOrDefaultAsync();
                    }
                }
```

**Comment by garayx** [comment]
Hi

You can use load multiple entities and pass the Keys collection:
https://ravendb.net/docs/article-page/5.4/csharp/client-api/session/loading-entities#load---multiple-entities
and then do the filtering on the `OtherField` on client side
